### PR TITLE
AA-220: Have a vertical check all leaf children

### PR DESCRIPTION
**Pull request by Dillon-Dumesnil**
We encountered an issue where a vertical was being marked complete
when it had a content library as a child because the content library
was being marked as complete and none of the actual problems in the
library were being checked. This now will check those children

**Merge checklist:**
- [x] All reviewers approved
- [x] CI build is green
- [x] Version bumped
- [x] Changelog record added
- [x] Documentation updated (not only docstrings)
- [x] Commits are squashed

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPi after tag-triggered build is 
      finished.
- [ ] Delete working branch (if not needed anymore)

### CODE REVIEW

File: completion/services.py [hunk 1]
```diff
@@ -76,6 +76,44 @@ def fill_in_run(block_key):
                 completions[candidate] = 0.0
         return completions
 
+    def get_selected_library_children(self, library_node):
```
Comment (cpennington) [inline completion/services.py:79]: I'm not wild about having to have explicit knowledge of how the library selects children. Does is there any common interface of library and regular blocks that just gets active children? I know if we were working on top of block structures, we could rely on the non-active children having already been filtered out

File: completion/services.py [hunk 2]
```diff
@@ -76,6 +76,44 @@ def fill_in_run(block_key):
...
+        Content Libraries will select a subset of the content in their library to
+        show to each learner. This function takes in a library content node and
+        returns the subset of its children that were selected for this user.
+        """
+        selected_children = []
+        all_children = library_node.get_children()
+        for block_type, block_id in library_node.selected:
+            selected_key = library_node.location.course_key.make_usage_key(block_type, block_id)
+            selected_children.extend(
+                list(filter(
```
Comment (cpennington) [inline completion/services.py:90]: Not: I tend to prefer list comprehensions for this sort of thing in python, because python lambdas are a bit wordy.

Less of a nit: `get_children` doesn potentially expensive work for all children, and then you filter them out. You could instwad iterate over `selected` and then load them from the runtime yourself.

File: completion/services.py [hunk 3]
```diff
@@ -76,6 +76,44 @@ def fill_in_run(block_key):
...
+                    lambda child, selected_key=selected_key: selected_key == child.scope_ids.usage_id, all_children
```
Comment (mikix) [inline completion/services.py:91]: You're passing in a new keyword argument to a lambda? Not sure I've seen that before, interesting. But couldn't you just refer to `selected_key` as in-scope even without the keyword arg?

File: completion/services.py [hunk 4]
```diff
@@ -92,7 +130,8 @@ def vertical_is_complete(self, item):
 
         # this is temporary local logic and will be removed when the whole course tree is included in completion
         child_locations = [
-            child.scope_ids.usage_id for child in item.get_children() if child.scope_ids.block_type != 'discussion'
+            child.scope_ids.usage_id for child in self.get_user_children(item)
+            if child.scope_ids.block_type != 'discussion'
```
Comment (mikix) [inline completion/services.py:134]: It seems potentially appropriate to move the `discussion` check inside `get_user_children`? Seems similar to the special handling for libraries.

Comment (Dillon-Dumesnil) [reply completion/services.py:134]: Yeah, I can do that. It was a remnant from the old version.

File: completion/services.py (hunk 3 above)
Comment (Dillon-Dumesnil) [reply completion/services.py:91]: Yeah, I was getting some pylint error about how selected_key changes value and how it's just taking the implied value from the scope. Googling around led me to passing it into the lambda as a kwarg and it fixed the pylint issue.

File: completion/services.py (hunk 2 above)
Comment (Dillon-Dumesnil) [reply completion/services.py:90]: Switched to list comprehension. Definitely cleaner.

Yeah, I wasn't sure about that for a  couple reasons:
1) The block transformer uses this similar format so I wasn't sure if it was a defensive measure in the situation of a selected block having been removed from the library and thus won't be in the children.
2) I'm not sure how to load from the runtime (modulestore?) (in general) and especially while in the completion service.

Let me know what you think.

File: completion/services.py (hunk 1 above)
Comment (Dillon-Dumesnil) [reply completion/services.py:79]: Yeah, I definitely do not care for it either. The only spot I found that is able to handle filtering out the unselected children was the library content block transformer, but we unfortunately can't use that here. This logic is largely based off what happens in that transformer (with some tweaks to work with the object type we have here). And I mentioned this same issue to Nimisha, but attempting to switch  to a block structure format was estimated at 2 devs for 4 weeks. So this is where I landed.

File: completion/services.py (hunk 2 above)
Comment (cpennington) [reply completion/services.py]: I think https://github.com/edx/edx-platform/blob/master/common/lib/xmodule/xmodule/library_content_module.py#L303-L309 could do what you want.

Comment (Dillon-Dumesnil) [reply completion/services.py]: Ah yes! There was a method in that file called `get_child_descriptors`that returns the list version of the `_get_selected_child_blocks`. Great call there, Cale!

File: completion/services.py [hunk 5]
```diff
@@ -76,6 +76,28 @@ def fill_in_run(block_key):
...
+    def get_user_children(self, node):
+        """
+        Verticals will sometimes have children that also have children (some
+        examples are content libraries, split tests, conditionals, etc.). This
+        function will recurse through the children to get to the bottom leaf
+        and return only those children. The nodes passed in should have already
+        taken the user into account so the proper blocks are shown for this user.
+        """
+        if node.scope_ids.block_type == 'library_content':
+            node_children = node.get_child_descriptors()
```
Comment (cpennington) [inline completion/services.py:90]: I think you can just use this method all the time, actually. You'll end up with the same thing as `get_children` in cases where the block doesn't have any dynamic children.

File: completion/services.py [hunk 6]
```diff
@@ -76,6 +76,28 @@ def fill_in_run(block_key):
                 completions[candidate] = 0.0
         return completions
 
+    def get_user_children(self, node):
```
Comment (cpennington) [inline completion/services.py:79]: I got confused by this method name, because you're actually getting `descendents` or maybe even `leaf_descendents`, since it's not just children being returned, it's doing the full leaf recursion.

File: completion/services.py (hunk 5 above)
Comment (cpennington) [reply completion/services.py:90]: Actually... so, I think `get_child_descriptors` only applies to `XModules`, not to pure `XBlocks`. But `get_children` is used by both. So, I'd use `get_child_descriptors` if it's available, and then `get_children` if it's not.

Comment (Dillon-Dumesnil) [reply completion/services.py:90]: Updated to prefer `get_child_descriptors` and fall back to `get_children`

File: completion/services.py (hunk 6 above)
Comment (Dillon-Dumesnil) [reply completion/services.py:79]: Yeah, I went with user children because it is really trying to get to the content relevant for a particular user. In the library content case, it wants the problems selected for that user. I try to call this out in the docstring as well, but perhaps I should add it higher in the description.

**Comment by Dillon-Dumesnil** [comment]
I think given that I'd be mocking out all of the information we would want to be testing for, we wouldn't be gaining much value from tests. As such, I'm going to opt out of writing tests for these changes and will be ready to revert/fix forward if we notice issues.

File: completion/services.py (hunk 6 above)
Comment (cpennington) [reply completion/services.py:79]: Maybe `get_user_leaf_blocks`? I'm not going to block on that, though.

File: completion/services.py [hunk 7]
```diff
@@ -76,6 +76,29 @@ def fill_in_run(block_key):
...
+        examples are content libraries, split tests, conditionals, etc.). This
+        function will recurse through the children to get to the bottom leaf
+        and return only those children.
+
+        Note: The nodes passed in should have already taken the user into
+        account so the proper blocks are shown for this user.
+        """
+        if node.scope_ids.block_type != 'discussion':
+            return []
+
```
Comment (xitij2000) [inline completion/services.py:91]: @Dillon-Dumesnil  I was hoping a refactoring here would get rid of this special casing and instead check if the block had [a `compeltion_mode` of `XBlockCompletionMode.EXCLUDED`](https://github.com/edx/edx-platform/blob/61e1eda20df2825a409db3e2d36c69d7c36d3e2d/openedx/core/lib/xblock_builtin/xblock_discussion/xblock_discussion/__init__.py#L43)

Currently, there is still a bug, in that blocks that are marked as excluded and will never have a completion score will still be counted, and any vertical with such a block will never get marked as completed. 
CC: @cpennington

Comment (xitij2000) [reply completion/services.py:91]: Sorry, I notice this has already been fixed. I came across this PR and figured it was the final say in the matter.

Comment (Dillon-Dumesnil) [reply completion/services.py:91]: @xitij2000 Ah yes, we realized the previous logic was not always doing as intended and Cale brought up using the EXCLUDED field. Thank you for also noticing and reaching out! This also reminded me that I forgot to update the CHANGELOG so thank you for that as well!
